pkg/util/purge/purge.go

```diff
@@ -0,0 +1,425 @@
+			return err
+		}
+		err = rc.cleanPrivateLink(ctx, resourceGroup)
+		if err != nil {
+			rc.log.Errorf("Cannot clean privatelinks: %s", err)
+			return err
+		}
+
+		if !rc.dryRun {
+			_, err := rc.resourcegroupscli.Delete(ctx, *resourceGroup.Name)
+			if err != nil {
+				rc.log.Errorf("Cannot delete resourceGroup: %s. Error: %w", *resourceGroup.Name, err)
+				return err
+			}
+		}
+		rc.log.Infof("deleting: %s", *resourceGroup.Name)
+	}
+
+	return nil
+}
+
+// cleanNetworking lists subnets in vnets and unnassign security groups
+func (rc *ResourceCleaner) cleanNetworking(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
+	vnets, err := rc.vnetscli.List(ctx, *resourceGroup.Name)
+	if err != nil {
+		rc.log.Errorf("%s: %s", *resourceGroup.Name, err)
+		return err
+	}
+
+	for _, vnet := range vnets {
+		var changed bool
+
+		for i := range *vnet.Subnets {
+			if !rc.dryRun {
+				(*vnet.Subnets)[i].NetworkSecurityGroup = nil
+				changed = true
+			} else {
+				rc.log.Infof("Removing NetworkSecurityGroup from vnet: %s, subnet: %s", *vnet.Name, *(*vnet.Subnets)[i].Name)
+				rc.log.Infof("updating vnet %s/%s", *resourceGroup.Name, *vnet.Name)
+			}
+		}
+
+		if changed {
+			rc.log.Printf("updating vnet %s/%s", *resourceGroup.Name, *vnet.Name)
+			_, err := rc.vnetscli.CreateOrUpdate(ctx, *resourceGroup.Name, *vnet.Name, vnet)
+			if err != nil {
+				rc.log.Errorf(err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+// cleanPrivateLink lists and unassigns all private links. If they are assigned the deletoin will fail
+func (rc *ResourceCleaner) cleanPrivateLink(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
+	plss, err := rc.privatelinkservicescli.List(ctx, *resourceGroup.Name)
+	if err != nil {
+		rc.log.Errorf("%s: %s", *resourceGroup.Name, err)
+		return err
+	}
+	for _, pls := range plss {
+		for _, peconn := range *pls.PrivateEndpointConnections {
+			rc.log.Infof("deleting private endpoint connection %s/%s/%s", *resourceGroup.Name, *pls.Name, *peconn.Name)
+			if !rc.dryRun {
+				_, err := rc.privatelinkservicescli.DeletePrivateEndpointConnection(ctx, *resourceGroup.Name, *pls.Name, *peconn.Name)
+				if err != nil {
+					rc.log.Errorf("Cannot delete privatelink: %s, %s", *pls.Name, err)
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// CleanAAD does the cleaning of the Active Directory
+//     - clean unsused app registration
+//     - clean unsused role assignments
+func (rc *ResourceCleaner) CleanAAD(ctx context.Context) error {
+
+	keepClientIDs, err := rc.listExistingClusters(ctx)
+	if err != nil {
+		rc.log.Errorf("Cannot list clusters to preserve: %w", err)
+		return err
+	}
+
+	// if the clientID was found in the database but the cluster resource group is scheduled for deletion.
+	// remove it from keep and delete it
+	for clientID := range rc.deleteClientIDs {
+		if _, ok := keepClientIDs[clientID]; ok {
+			delete(keepClientIDs, clientID)
+		}
+	}
```

Comment:
I might be misreading this.
I think if cluster document is in the database, we should just keepit. This might simplify tagging for "ToKeep" for users who wants to keep cluster for debugging.